packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java

```diff
@@ -0,0 +1,148 @@
+    try {
+      PackageInfo info = packageManager.getPackageInfo(packageName, 0);
+      promise.resolve((double)info.firstInstallTime);
+    } catch (PackageManager.NameNotFoundException e) {
+      Log.e(TAG, "Exception: ", e);
+      promise.reject("ERR_APPLICATION_PACKAGE_NAME_NOT_FOUND", "Unable to get install time of this application. Could not get package info or package name.", e);
+    }
+  }
+
+  @ExpoMethod
+  public void getLastUpdateTimeAsync(Promise promise) {
+    PackageManager packageManager = mContext.getPackageManager();
+    String packageName = mContext.getPackageName();
+    try {
+      PackageInfo info = packageManager.getPackageInfo(packageName, 0);
+      promise.resolve((double)info.lastUpdateTime);
+    } catch (PackageManager.NameNotFoundException e) {
+      Log.e(TAG, "Exception: ", e);
+      promise.reject("ERR_APPLICATION_PACKAGE_NAME_NOT_FOUND", "Unable to get last update time of this application. Could not get package info or package name.", e);
+    }
+  }
+
+
+  @ExpoMethod
+  public void getInstallReferrerAsync(final Promise promise) {
+    final StringBuilder installReferrer = new StringBuilder();
+
+    final InstallReferrerClient referrerClient;
+    referrerClient = InstallReferrerClient.newBuilder(mContext).build();
+    referrerClient.startConnection(new InstallReferrerStateListener() {
+      @Override
+      public void onInstallReferrerSetupFinished(int responseCode) {
+        switch (responseCode) {
+          case InstallReferrerClient.InstallReferrerResponse.OK:
+            // Connection established and response received
+            try {
+              ReferrerDetails response = referrerClient.getInstallReferrer();
+              installReferrer.append(response.getInstallReferrer());
+            } catch (RemoteException e) {
+              Log.e(TAG, "Exception: ", e);
+              promise.reject("ERR_APPLICATION_INSTALL_REFERRER_REMOTE_EXCEPTION", "RemoteException getting install referrer information. This may happen if the process hosting the remote object is no longer available.", e);
+            }
+            promise.resolve(installReferrer.toString());
+            break;
+          case InstallReferrerClient.InstallReferrerResponse.FEATURE_NOT_SUPPORTED:
+            // API not available in the current Play Store app
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER_UNAVAILABLE", "The current Play Store app doesn't provide the installation referrer API, or the Play Store may not be installed.");
+            break;
+          case InstallReferrerClient.InstallReferrerResponse.SERVICE_UNAVAILABLE:
+            // Connection could not be established
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER_CONNECTION", "Could not establish a connection to Google Play");
+            break;
+          default:
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "General error: response code " + responseCode);
```

Comment:
Seeing "General error: response code XX" in your error logs without context is going to be hard to debug -- add more context:

```suggestion
            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "General error retrieving the install referrer: response code " + responseCode);
```